provider/digital_ocean.go

```diff
@@ -0,0 +1,279 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"golang.org/x/oauth2"
+
+	"github.com/digitalocean/godo"
+	"github.com/digitalocean/godo/context"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+)
+
+// DigitalOceanProvider is an implementation of Provider for Digital Ocean's DNS.
+type DigitalOceanProvider struct {
+	Client godo.DomainsService
+	// only consider hosted zones managing domains ending in this suffix
+	domainFilter string
+	DryRun       bool
+}
+
+// TokenSource inherits AccessToken to initialize a new Digital Ocean client.
+type TokenSource struct {
+	AccessToken string
+}
+
+// DigitalOceanChange differentiates between ChangActions
+type DigitalOceanChange struct {
+	Action            string
+	ResourceRecordSet godo.DomainRecord
+}
+
+const (
+	// DigitalOceanCreate is a ChangeAction enum value
+	DigitalOceanCreate = "CREATE"
+	// DigitalOceanDelete is a ChangeAction enum value
+	DigitalOceanDelete = "DELETE"
+	// DigitalOceanUpdate is a ChangeAction enum value
+	DigitalOceanUpdate = "UPDATE"
+)
+
+// Token returns oauth2 token struct in order to use Digital Ocean's API.
+func (t *TokenSource) Token() (*oauth2.Token, error) {
+	token := &oauth2.Token{
+		AccessToken: t.AccessToken,
+	}
+	return token, nil
+}
+
+// NewDigitalOceanProvider initializes a new DigitalOcean DNS based Provider.
+func NewDigitalOceanProvider(domainFilter string, dryRun bool) (Provider, error) {
+	token := os.Getenv("DO_TOKEN")
+	tokenSource := &TokenSource{
+		AccessToken: token,
+	}
+	if len(token) == 0 {
+		return nil, fmt.Errorf("No token found")
+	}
+	oauthClient := oauth2.NewClient(oauth2.NoContext, tokenSource)
+	client := godo.NewClient(oauthClient)
+
+	provider := &DigitalOceanProvider{
+		Client:       client.Domains,
+		domainFilter: domainFilter,
+		DryRun:       dryRun,
+	}
+	return provider, nil
+}
+
+// Zones returns the list of hosted zones.
+func (p *DigitalOceanProvider) Zones() ([]godo.Domain, error) {
+	result := []godo.Domain{}
+
+	zones, _, err := p.Client.List(context.TODO(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range zones {
+		if strings.HasSuffix(zone.Name, p.domainFilter) {
+			result = append(result, zone)
+		}
+	}
+
+	return result, nil
+}
+
+// Records returns the list of records in a given zone.
+func (p *DigitalOceanProvider) Records() ([]*endpoint.Endpoint, error) {
+	zones, err := p.Zones()
+	if err != nil {
+		return nil, err
+	}
+	endpoints := []*endpoint.Endpoint{}
+	for _, zone := range zones {
+		records, _, err := p.Client.Records(context.TODO(), zone.Name, nil)
```

Comment:
Here's an example of pagination: https://github.com/digitalocean/godo#pagination